Sources/TuistCore/ValueGraph/ValueGraph.swift

```diff
@@ -16,15 +22,114 @@ public struct ValueGraph: Equatable {
         self.projects = projects
         self.packages = packages
         self.targets = targets
         self.dependencies = dependencies
     }
+
+    init(name: String, path: AbsolutePath, cache: GraphLoaderCaching) {
+        self.name = name
+        self.path = path
+        projects = cache.projects
+        packages = cache.packages.mapValues { packages in
+            packages.reduce(into: [String: Package]()) { $0[$1.name] = $1.package }
+        }
+        targets = cache.targetNodes.mapValues { targets in
+            targets.mapValues { $0.target }
+        }
+        dependencies = ValueGraph.dependencies(from: cache)
+    }
+
+    init(name: String, path: AbsolutePath, graph: Graph) {
+        self.name = name
+        self.path = path
+        projects = graph.projects.reduce(into: [AbsolutePath: Project]()) { $0[$1.path] = $1 }
+        packages = graph.packages.reduce(into: [AbsolutePath: [String: Package]]()) { acc, package in
+            var packages = acc[package.path, default: [:]]
+            packages[package.name] = package.package
+            acc[package.path] = packages
+        }
+        targets = graph.targets.mapValues { targets in targets.reduce(into: [String: Target]()) { $0[$1.name] = $1.target } }
+        dependencies = ValueGraph.dependencies(from: graph)
+    }
+
+    /// Given a graph loader cache, it returns a dictionary representing the dependency tree.
+    /// - Parameter cache: Cache generated after loading the projects.
+    /// - Returns: Dependency tree.
+    private static func dependencies(from graph: Graph) -> [ValueGraphDependency: Set<ValueGraphDependency>] {
+        var dependencies: [ValueGraphDependency: Set<ValueGraphDependency>] = [:]
+        graph.forEach { map(node: $0, into: &dependencies) }
+        return dependencies
+    }
+
+    /// Given a graph loader cache, it returns a dictionary representing the dependency tree.
+    /// - Parameter cache: Cache generated after loading the projects.
+    /// - Returns: Dependency tree.
+    private static func dependencies(from cache: GraphLoaderCaching) -> [ValueGraphDependency: Set<ValueGraphDependency>] {
+        var dependencies: [ValueGraphDependency: Set<ValueGraphDependency>] = [:]
+        cache.forEach { map(node: $0, into: &dependencies) }
+        return dependencies
+    }
+
+    /// Traverses the node and its dependencies and adds them to the dependencies dictionary.
+    /// - Parameters:
+    ///   - node: Node whose dependencies will be mapped.
+    ///   - dependencies: Dictionary containing the dependency tree.
+    private static func map(node: GraphNode, into dependencies: inout [ValueGraphDependency: Set<ValueGraphDependency>]) {
+        let nodeDependency = dependency(from: node)
+        var nodeDependencies: Set<ValueGraphDependency>! = dependencies[nodeDependency]
+        if nodeDependencies == nil {
+            nodeDependencies = Set()
+        }
+
+        if let targetNode = node as? TargetNode {
```

Comment:
minor suggestion:  You can leverage `switch` with type casting in many of those `if let as?`where we're dealing with multiple cases  _(not sure if its any better)_

e.g.

```swift
switch node {
   case let target as TargetNode:
       // ...
   case let frameworkNode as FrameworkNode:
       // ...
   case let xcframeworkNode as XCFrameworkNode:
      // ...
   default:
         break
 }
```